### PR TITLE
[McByte part 3] Add SAM box mask generator

### PR DESCRIPTION
**Pull request by tstanczyk95**
## What does this PR do?

This PR adds a SAM-based MaskGenerator implementation for McByte.

Included:
- SAMBoxMaskGenerator
- automatic checkpoint download
- mask conversion utilities
- unit test
- manual visualization script

Not included:
- Cutie integration
- mask propagation
- MaskManager lifecycle logic

## Type of Change

- New feature (non-breaking change that adds functionality)

## Testing

- [x] I have tested this change locally
- [x] I have added/updated tests for this change

**Test details:**
- Added a unit test for segmentation mask correct format conversion and shape maintaining.
- Added a visual test script for creating masks and overlying them on the image.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [ ] [McByte as a whole is still in the development phase] I have updated the documentation accordingly (if applicable)

## Additional Context

It requires installation of torch, torch vision and SAM as specified in the [SAM GitHub repo](https://github.com/facebookresearch/segment-anything#installation).

Tested on CPU and on GPU (CUDA 12.6) - separate Anaconda environments.

CPU extra parts installation:
`pip install torch torch vision`
`pip install git+https://github.com/facebookresearch/segment-anything.git`

GPU extra parts installation (assuming that CUDA is installed and enabled):
`pip3 install torch torchvision --index-url https://download.pytorch.org/whl/cu126 # taken from: https://pytorch.org/get-started/locally/`
`pip install git+https://github.com/facebookresearch/segment-anything.git # same as for CPU installation`

Visual test performed on [SoccerNet-tracking](https://www.soccer-net.org/tasks/tracking):
- sequence SNMOT-116
- frame 000548.jpg
- oracle detections converted to xyxy format

Example test run for the visual test (you can also run it just with a single bounding box):

```
python visual_tests/visualize_sam_mask_generator.py \
  --image-path your_dataset_path/SoccerNet/tracking/test/SNMOT-116/img1/000548.jpg \
  --box 177,482,249,631 \
  --box 1241,569,1297,716 \
  --box 191,344,240,459 \
  --box 686,298,751,426 \
  --box 625,469,686,623 \
  --box 1382,378,1440,497 \
  --box 1220,480,1240,498 \
  --output-path visual_tests/outputs/sam_mask_generator_output.jpg \
  --device cpu
```

When `SAMBoxMaskGenerator` is instantiated, it checks if the pre-trained SAM model is already saved in the pointed location. If the model not saved, then it downloaded. Default location for the pre-trained model: `trackers_root/models/sam/sam_vit_b_01ec64.pth`. Model from [here](https://github.com/facebookresearch/segment-anything#model-checkpoints), model taken from the official [SAM GitHub repo](https://github.com/facebookresearch/segment-anything#model-checkpoints): `vit_b`.

### CODE REVIEW

File: src/trackers/core/mcbyte/masks/sam.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+logger = logging.getLogger(__name__)
+
+SAM_CHECKPOINT_URLS = {
+    "vit_b": "https://dl.fbaipublicfiles.com/segment_anything/sam_vit_b_01ec64.pth",
+}
+
+SAM_DEFAULT_CHECKPOINT_PATHS = {
+    "vit_b": Path("models/sam/sam_vit_b_01ec64.pth"),
+}
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/sam.py:32]: SAM model is used within SAMBoxMaskGenerator which is designed and serves as an internal component of McByte. It is an internal repository-local default for development and testing. Users can override it through `checkpoint_path`. We can revisit a cache-based approach if SAM (and Cutie later) become permanent first-class dependencies.

File: src/trackers/core/mcbyte/masks/sam.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        Args:
+            frame: Current RGB frame with shape ``(H, W, 3)``.
+            tracklets: Tracklet snapshots containing tracker IDs and ``xyxy``
+                bounding boxes.
+
+        Returns:
+            MaskOutput containing masks with shape ``(N, H, W)``, where ``N`` is
+            the number of input tracklets. ``tracklet_mask_dict`` maps each
+            tracker ID to its local mask index in the returned mask array.
+        """
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/sam.py:129]: Addressed in d446e0c.

File: src/trackers/core/mcbyte/masks/sam.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            msg = (
+                "SAM support requires Segment Anything. "
+                "Install it via `pip install "
+                "git+https://github.com/facebookresearch/segment-anything.git`."
+            )
+            raise ImportError(msg) from exc
+
+        self.checkpoint_path = (
+            Path(checkpoint_path) if checkpoint_path is not None else SAM_DEFAULT_CHECKPOINT_PATHS[model_type]
+        )
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/sam.py]: Addressed in d446e0c.

File: src/trackers/core/mcbyte/masks/sam.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        raise ValueError(f"No default checkpoint URL for model_type={model_type!r}.")
+
+    parsed_url = urlparse(checkpoint_url)
+    if parsed_url.scheme != "https":
+        raise ValueError(f"Unsupported checkpoint URL scheme: {parsed_url.scheme!r}")
+
+    checkpoint_path.parent.mkdir(parents=True, exist_ok=True)
+
+    logger.info("Downloading SAM checkpoint to %s", checkpoint_path)
+    urlretrieve(checkpoint_url, checkpoint_path)  # noqa: S310
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/sam.py]: Addressed in d446e0c.
Using `trackers.utils.downloader._download_file()`.

File: tests/core/test_mcbyte_sam_mask_generator.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# ------------------------------------------------------------------------
+# Trackers
+# Copyright (c) 2026 Roboflow. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 [see LICENSE for details]
+# ------------------------------------------------------------------------
+
+from __future__ import annotations
+
+import torch
+
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply tests/core/test_mcbyte_sam_mask_generator.py]: Addressed in d446e0c.

File: visual_tests/visualize_sam_mask_generator.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    generator = SAMBoxMaskGenerator(
+        model_type=args.model_type,
+        device=validate_device(args.device),
+    )
+
+    print(f"Generator device: {generator.device}")
+    if generator.device.type == "cuda":
+        print(f"GPU: {torch.cuda.get_device_name(generator.device)} (CUDA {torch.version.cuda})")
+    else:
+        print("GPU: CPU")
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply visual_tests/visualize_sam_mask_generator.py]: Addressed in d446e0c.

**Comment by tstanczyk95** [comment]
@Borda do you think it's good for merging into `dev/mcbyte`?

**Comment by Borda** [comment]
@tstanczyk95 if you feel confident that your reply to the copilot comment is sufficient, or disagree with his claim, or you applied the suggestion, pls make it as resolved :)

File: src/trackers/core/mcbyte/masks/sam.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+logger = logging.getLogger(__name__)
+
+SAM_CHECKPOINT_URLS = {
+    "vit_b": "https://dl.fbaipublicfiles.com/segment_anything/sam_vit_b_01ec64.pth",
+}
+
+SAM_DEFAULT_CHECKPOINT_PATHS = {
+    "vit_b": Path("models/sam/sam_vit_b_01ec64.pth"),
+}
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/sam.py:32]: Why is the default a dict and just a single path/str?

File: src/trackers/core/mcbyte/masks/sam.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        model_type: str = "vit_b",
+        device: str = "cpu",
+    ) -> None:
+        try:
+            from segment_anything import SamPredictor, sam_model_registry
+        except ImportError as exc:
+            msg = (
+                "SAM support requires Segment Anything. "
+                "Install it via `pip install "
+                "git+https://github.com/facebookresearch/segment-anything.git`."
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/sam.py:91]: I would suggest having it as extras, but since it's just a URL, it's not allowed by pypi, so a bit challenging...

File: visual_tests/visualize_sam_mask_generator.py
Comment (Borda) [inline visual_tests/visualize_sam_mask_generator.py:1]: Is this temporary?

File: src/trackers/core/mcbyte/masks/sam.py (hunk 7 above)
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/sam.py:32]: The mapping is intentionally kept as a dictionary because the constructor already supports model_type, and additional SAM variants (e.g. `vit_l`, `vit_h`) can be added later without changing the API or lookup logic. At the moment only `vit_b` is registered because that is the model used by McByte. 

Added an extra comment to make it clear in f42f742 .

File: src/trackers/core/mcbyte/masks/sam.py (hunk 8 above)
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/sam.py:91]: Extras would indeed be preferable but are constrained by how SAM is distributed.

File: visual_tests/visualize_sam_mask_generator.py
Comment (tstanczyk95) [reply visual_tests/visualize_sam_mask_generator.py:1]: At the moment, I would prefer to keep this script. It is intended as a development and debugging utility rather than part of the production tracking pipeline. It has already proven useful for visually validating intermediate results, reproducing issues, and verifying changes during implementation, also in the subsequent PRs. We can revisit whether to keep or remove it when preparing the project for a final release.

I slightly updated its docstring in f42f742 to emphasize that.
